File: app/src/main/java/com/drawable/Rectangle.java
package com.drawable;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;

import com.util.Color;

@SuppressWarnings({"unused", "WeakerAccess"})
public class Rectangle extends Drawable implements Shape {
    private int width;
    private int height;

    public Rectangle(double x, double y, int width, int height, int color) {
        this(x, y, width, height, color, 0, 0);
    }

    public Rectangle(double x, double y, int width, int height, Color color) {
        this(x, y, width, height, color.getInt(), 0, 0);
    }

    public Rectangle(double x, double y, int width, int height, int color, double xVelocity, double yVelocity) {
        super(x, y, color, xVelocity, yVelocity);
        this.width = width;
        this.height = height;
    }

    public Rectangle(double x, double y, int width, int height, Color color, double xVelocity, double yVelocity) {
        this(x, y, width, height, color.getInt(), xVelocity, yVelocity);
    }

    public Rectangle(Rectangle r) {
        super(r);
        this.width = r.getWidth();
        this.height = r.getHeight();
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public int getSideLength() {
        if (!(this instanceof Square)) return 0;
        return width;
    }

    public void setSideLength(int sideLength) {
        if (!(this instanceof Square)) return;
        width = sideLength;
        height = sideLength;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    public void setHeight(int height) {
        this.height = height;
    }

    @Override
    public double area() {
        return width * height;
    }

    @Override
    public double perimeter() {
        return (width * 2) + (height * 2);
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Rectangle)) return false;
        Rectangle other = (Rectangle) o;
        return this.width == other.width && this.height == other.height && this.getColor() == other.getColor();
    }

    @Override
    public void draw(Canvas canvas, Paint paint) {
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(getColor());

        RectF rect = new RectF((float) getX(), (float) getY(), (float) (getX() + width), (float) (getY() + height));
        canvas.drawRect(rect, paint);
        animate();
    }

    @Override
    public String toString() {
        if (this instanceof Square) {
            return super.toString() + "\n \tSide length: " +
                    getSideLength() + "\n \tArea: " + area() + "\n \tPerimeter: " + perimeter() + "\n";
        }
        return super.toString() + "\n \tWidth: " + width + "\n \tHeight: "+ height + "\n \tColor: " +
                getColorAsHex() + "\n \tArea: " + area() + "\n \tPerimeter: " + perimeter() + "\n";
    }
}

File: app/src/main/java/com/drawable/Rectangle.kt
package com.drawable

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF

import com.util.Color

@Suppress("unused", "NAME_SHADOWING", "EqualsOrHashCode")
open class Rectangle : Drawable, Shape {
    var width: Int = 0
    var height: Int = 0

    private var sideLength: Int
        get() = (this as? Square)?.width ?: 0
        set(sideLength) {
            if (this !is Square) return
            width = sideLength
            height = sideLength
        }

    constructor(x: Double, y: Double, width: Int, height: Int, color: Color) : this(x, y, width, height, color.int, 0.0, 0.0)

    constructor(x: Double, y: Double, width: Int, height: Int, color: Int, xVelocity: Double = 0.0, yVelocity: Double = 0.0) :
            super(x, y, color, xVelocity, yVelocity) {
        this.width = width
        this.height = height
    }

    constructor(x: Double, y: Double, width: Int, height: Int, color: Color, xVelocity: Double, yVelocity: Double) :
            this(x, y, width, height, color.int, xVelocity, yVelocity)

    constructor(r: Rectangle) : super(r) {
        this.width = r.width
        this.height = r.height
    }

    override fun area(): Double {
        return (width * height).toDouble()
    }

    override fun perimeter(): Double {
        return (width * 2 + height * 2).toDouble()
    }

    override fun equals(other: Any?): Boolean {
        if (other !is Rectangle) return false
        val other = other as Rectangle?
        return this.width == other!!.width && this.height == other.height && this.color == other.color
    }

    override fun draw(canvas: Canvas, paint: Paint) {
        paint.style = Paint.Style.FILL
        paint.color = color

        val rect = RectF(x.toFloat(), y.toFloat(), (x + width).toFloat(), (y + height).toFloat())
        canvas.drawRect(rect, paint)
        animate()
    }

    override fun toString(): String {
        return if (this is Square) {
            super.toString() + "\n \tSide length: " +
                    sideLength + "\n \tArea: " + area() + "\n \tPerimeter: " + perimeter() + "\n"
        } else super.toString() + "\n \tWidth: " + width + "\n \tHeight: " + height +
                "\n \tArea: " + area() + "\n \tPerimeter: " + perimeter() + "\n"
    }
}